open-apparel-registry/open-apparel-registry
# Update Django to 3.2

## Overview

Updates Django to 3.2 & Python to 3.7. Also updates a number of Django-related libraries, as well as a few libraries that required updating to meet the new Python version.

I did _not_ update `dedupe`, which is why we're on Python 3.7 (the latest version supported by the version of `dedupe` we're using). We'll have to defer further Python updates until we update `dedupe`.

Connects #1908 

## Notes

My process was as follows:
 - I read through the Django release notes and checked for any backwards incompatible changes
 - I updated the underlying Dockerfile to use the latest versions of Python & Django (I later backdated the Python version)
 - I went through the CHANGELOG for each requirement in our `requirements.txt` to see if they needed an update based on either the new Django version or the new Python version (this mostly matched Tai's investigation but I spotted a few more that needed updating)
 - `scripts/update`
 - I added `-W all` to the call to `python manage.py` in `scripts/manage`, and then looked for warnings / errors in:
   - `scripts/manage check`
   - `scripts/manage test`
 - Similarly I added `entrypoint: "python -Wa /usr/local/bin/gunicorn"` under the `django` container in the `docker--compose.yml` file and ran `scripts/server`
 - You can ignore warnings from specific modules, so after deciding it was OK to ignore and/or infeasible to fix warnings from these, I got to 0 warnings w/ these arguments: `-Wa -W ignore:::thefuzz.fuzz-W ignore:::fastcluster -W ignore:::dedupe.blocking -W ignore:::dedupe.core -W ignore:::rest_auth.urls -W ignore:::rest_auth.registration.urls`
   - `thefuzz` & `fastcluster` are both related to `dedupe`, which I mentioned above.
   - the `rest_auth` warnings aren't fatal (yet) so I deferred them since we'll likely have to migrate away from the library entirely

In the end my code changes to check for warnings ended up like this:
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 2402c0c..1dfa757 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -37,6 +37,7 @@ services:
     depends_on:
       database:
         condition: service_healthy
+    entrypoint: "python -Wa -W ignore:::fastcluster -W ignore:::dedupe.blocking -W ignore:::rest_auth.urls -W ignore:::rest_auth.registration.urls /usr/local/bin/gunicorn"
     command:
       - "-b :8081"
       - "--reload"
diff --git a/scripts/manage b/scripts/manage
index f0ccd7d..855bb26 100755
--- a/scripts/manage
+++ b/scripts/manage
@@ -20,6 +20,6 @@ if [ "${BASH_SOURCE[0]}" = "${0}" ]; then
         docker-compose \
             run --rm --entrypoint python \
             django \
-            manage.py "$@"
+            -Wa -W ignore:::fastcluster -W ignore:::dedupe.blocking -W ignore:::dedupe.core -W ignore:::rest_auth.urls -W ignore:::rest_auth.registration.urls -W ignore:::thefuzz.fuzz manage.py "$@"
     fi
 fi
```

## Testing Instructions

* `scripts/update`
* Run the application and look for any bugs
* I don't know if we have code coverage set up on this project, but if so it might be good to look to that to see what is not covered by the test suite and focus on that for manual testing

## Checklist
